src/View/CastingService.php

```diff
@@ -0,0 +1,100 @@
+<?php
+
+namespace SilverStripe\View;
+
+use LogicException;
+use SilverStripe\Core\ClassInfo;
+use SilverStripe\Core\Injector\Injectable;
+use SilverStripe\Core\Injector\Injector;
+use SilverStripe\Model\ArrayData;
+use SilverStripe\Model\List\ArrayList;
+use SilverStripe\Model\ModelData;
+use SilverStripe\ORM\FieldType\DBBoolean;
+use SilverStripe\ORM\FieldType\DBFloat;
+use SilverStripe\ORM\FieldType\DBInt;
+use SilverStripe\ORM\FieldType\DBText;
+
+class CastingService
+{
+    use Injectable;
+
+    /**
+     * Cast a value to the relevant object (usually a DBField instance) for use in the view layer.
+     *
+     * @param null|array|ModelData $source Where the data originates from. This is used both to check for casting helpers
+     * and to help set the value in cast DBField instances.
+     * @param bool $strict If true, an object will be returned even if $data is null.
+     */
+    public function cast(mixed $data, null|array|ModelData $source = null, string $fieldName = '', bool $strict = false): ?object
+    {
+        // null is null - we shouldn't cast it to an object, because that makes it harder
+        // for downstream checks to know there's "no value".
+        if (!$strict && $data === null) {
+            return null;
+        }
+
+        // Assume anything that's an object is intentionally using whatever class it's using
+        // and don't cast it.
+        if (is_object($data)) {
+            return $data;
+        }
+
+        $service = null;
+        if ($source instanceof ModelData) {
+            $service = $source->castingHelper($fieldName);
+        }
+
+        // Cast to object if there's an explicit casting for this field
+        // Explicit casts take precedence over array casting
+        if ($service) {
+            $castObject = Injector::inst()->create($service, $fieldName);
+            if (!ClassInfo::hasMethod($castObject, 'setValue')) {
```

Comment:
Will we every cast to something that's not a DBField? Should we just check if is_a() DBField instead?

Throwing this exception should probably live inside `castingHelper()` not here

Reply:
> Will we every cast to something that's not a DBField?

We already do cast to `ArrayList` and `ArrayData` - and there's nothing stopping `ModelData.casting` config from including any injectable service.

We can tighten that up later as a separate effort if you want to, but I think this PR is complex enough without the additional effort to find out exactly what might break by being strict here.

Reply:
> PR is complex enough without the additional effort to find out exactly what might break by being strict here.

But .. this is new code? Or was this existing code copied in from another existing class?

Reply:
Most of the code in this class is just lifted from `ModelData::castingHelper()` and the related methods in `ModelData`.

Reply:
OK all good, just leave as is then